ddl/table_lock.go

```diff
@@ -0,0 +1,222 @@
+			return i
+		}
+	}
+	return -1
+}
+
+// lockTable uses to check table locked and acquire the table lock for the request session.
+func lockTable(tbInfo *model.TableInfo, idx int, arg *lockTablesArg) error {
+	if !tbInfo.IsLocked() {
+		tbInfo.Lock = &model.TableLockInfo{
+			Tp: arg.LockTables[idx].Tp,
+		}
+		tbInfo.Lock.Sessions = append(tbInfo.Lock.Sessions, arg.SessionInfo)
+		return nil
+	}
+	if tbInfo.Lock.State == model.TableLockStatePreLock {
+		return nil
+	}
+	if tbInfo.Lock.Tp == model.TableLockRead && arg.LockTables[idx].Tp == model.TableLockRead {
+		sessionIndex := findSessionInfoIndex(tbInfo.Lock.Sessions, arg.SessionInfo)
+		// repeat lock.
+		if sessionIndex >= 0 {
+			return nil
+		}
+		tbInfo.Lock.Sessions = append(tbInfo.Lock.Sessions, arg.SessionInfo)
+		return nil
+	}
+
+	// Unlock tables should execute before lock tables.
+	// Normally execute to here is impossible.
+	return infoschema.ErrTableLocked.GenWithStackByArgs(tbInfo.Name.L, tbInfo.Lock.Tp, tbInfo.Lock.Sessions[0])
+}
+
+// checkTableLocked uses to check whether table was locked.
+func checkTableLocked(tbInfo *model.TableInfo, lockTp model.TableLockType, sessionInfo model.SessionInfo) error {
+	if !tbInfo.IsLocked() {
+		return nil
+	}
+	if tbInfo.Lock.State == model.TableLockStatePreLock {
+		return nil
+	}
+	if tbInfo.Lock.Tp == model.TableLockRead && lockTp == model.TableLockRead {
+		return nil
+	}
+	sessionIndex := findSessionInfoIndex(tbInfo.Lock.Sessions, sessionInfo)
+	// If the request session already locked the table before, In other words, repeat lock.
+	if sessionIndex >= 0 {
+		if tbInfo.Lock.Tp == lockTp {
+			return nil
+		}
+		// If no other session locked this table.
+		if len(tbInfo.Lock.Sessions) == 1 {
+			return nil
+		}
+	}
+	return infoschema.ErrTableLocked.GenWithStackByArgs(tbInfo.Name.L, tbInfo.Lock.Tp, tbInfo.Lock.Sessions[0])
+}
+
+// unlockTables uses unlock a batch of table lock one by one.
+func unlockTables(t *meta.Meta, job *model.Job, arg *lockTablesArg) (ver int64, err error) {
+	if arg.IndexOfUnlock >= len(arg.UnlockTables) {
+		return ver, nil
+	}
+	job.SchemaID = arg.UnlockTables[arg.IndexOfUnlock].SchemaID
+	job.TableID = arg.UnlockTables[arg.IndexOfUnlock].TableID
+	tbInfo, err := getTableInfo(t, job.TableID, job.SchemaID)
+	if err != nil {
+		if infoschema.ErrDatabaseNotExists.Equal(err) || infoschema.ErrTableNotExists.Equal(err) {
+			// The table maybe has been dropped. just ignore this err and go on.
+			arg.IndexOfUnlock++
+			job.Args = []interface{}{arg}
+			return ver, nil
+		} else {
+			return ver, err
+		}
+	}
+
+	ver, err = unlockTableAndUpdate(t, job, tbInfo, arg)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	arg.IndexOfUnlock++
+	job.Args = []interface{}{arg}
+	return ver, nil
+}
+
+// unlockTableAndUpdate uses to unlock table lock that hold by the session.
+func unlockTableAndUpdate(t *meta.Meta, job *model.Job, tbInfo *model.TableInfo, arg *lockTablesArg) (ver int64, err error) {
```

Comment:
why not return a boolean to indicate whether the unlock is succeeded, then we can update the schema version.

Never try to do more than one thing in one func.